packages/backend/src/core/discovery/source/SourceCodeService.ts

```diff
@@ -0,0 +1,47 @@
+import { EthereumAddress } from '@l2beat/shared'
+import { zip } from 'lodash'
+
+import { DiscoveryProvider } from '../provider/DiscoveryProvider'
+import { deduplicateAbi } from './deduplicateAbi'
+import { getDerivedName } from './getDerivedName'
+import { processSources } from './processSources'
+
+export interface ContractSources {
+  name: string
+  isVerified: boolean
+  abi: string[]
+  abis: Record<string, string[]>
+  files: Record<string, string>[]
+}
+
+export class SourceCodeService {
+  constructor(private readonly provider: DiscoveryProvider) {}
+
+  async getSources(
+    address: EthereumAddress,
+    implementations?: EthereumAddress[],
+  ): Promise<ContractSources> {
+    const addresses = [address, ...(implementations ?? [])]
+    const metadata = await Promise.all(
+      addresses.map((x) => this.provider.getMetadata(x)),
+    )
+
+    const name = getDerivedName(metadata.map((x) => x.name))
```

Comment:
Yes, our bot will detect a lot of changes. @sz-piotr please restore the previous functionality. Generally let's keep this refactoring as close to the current logic as possible. Potential feature updates can take place in the future, and it will be much easier with refactored code.

Reply:
Agreed